File: ruby/lesson 12/lru_cache.rb
# https://leetcode.com/problems/lru-cache/
    
class LRUCache
    def initialize(capacity)
        @capacity = capacity
        @map = {}
        @list = DoubleLinkedList.new 
    end

    def get(key)
        return -1 unless @map.include?(key)

        node = @map[key]
        @list.move_to_head(node)
        node.val
    end

    def put(key, value)
        if @map.include?(key)
            node = @map[key]
            node.val = value
            @list.move_to_head(node)
        else
            node = Node.new(key, value)
            @map[key] = node
            @list.add_to_head(node)
            
            @map.delete(@list.remove_from_tail.key) if @map.size > @capacity
        end
    end
end

class Node
    attr_accessor :key, :val, :prev, :next

    def initialize(key, value)
        @key = key
        @val = value
        @prev = nil
        @next = nil
    end
end

class DoubleLinkedList
    def initialize()
        @head = Node.new(0, 0)
        @tail = Node.new(0, 0)
        @head.prev = @tail
        @tail.next = @head
    end
    
    def add_to_head(node)
        prev = @head.prev
        prev.next = node
        node.prev = prev
        node.next = @head
        @head.prev = node
        node
    end

    def move_to_head(node)
        prev = node.prev
        _next = node.next
        prev.next = _next
        _next.prev = prev
        add_to_head(node)
    end

    def remove_from_tail()
        result = @tail.next
        @tail.next = result.next
        result.next.prev = @tail
        result.prev = nil
        result.next = nil
        result
    end
end

File: javascript/lesson 12/lruCache.js
// https://leetcode.com/problems/lru-cache/

class LRUCache {
    constructor(capacity) {
        this.capacity = capacity;
        this.map = new Map();
        this.list = new DoubleLinkedList();
    }
    
    get(key) {
        if (this.map.has(key)) {
            const node = this.map.get(key);
            this.list.moveToHead(node);
            return node.val;        
        }
        return -1;        
    }
    
    put(key, value) {
        if (this.map.has(key)) {
            const node = this.map.get(key);
            node.val = value;
            this.list.moveToHead(node);
        } else {
            const node = new Node(key, value);
            this.map.set(key, node);
            this.list.addToHead(node);
            if (this.map.size > this.capacity) {
                const r = this.list.removeFromTail();
                this.map.delete(r.key);
            }
        }
    }
    
}

class Node {
    constructor(key, value) {
        this.key = key;
        this.val = value;
        this.prev = null;
        this.next = null;
    }
    
}

class DoubleLinkedList {
    constructor() {
        this.head = new Node();
        this.tail = new Node();
        this.head.prev = this.tail;
        this.tail.next = this.head;
    }
    
    addToHead(node) {
        const prev = this.head.prev;            
        prev.next = node;
        node.prev = prev;
        node.next = this.head;
        this.head.prev = node;
        return node;
    }

    moveToHead(node) {
        const prev = node.prev;
        const next = node.next;            
        prev.next = next;
        next.prev = prev;
        this.addToHead(node);
    }

    removeFromTail() {
        const result = this.tail.next;            
        this.tail.next = result.next;
        result.next.prev = this.tail;
        result.prev = null;
        result.next = null;
        return result;
    }
}
